cleanup localscope implementtation (`RelativeName` and `ExtRelativeName`)

goto40 opened the pull request:
<!-- Please don't remove/change code review checklist -->

## Code review checklist

- [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [x] Title summarizes what is changing
- [x] Commit messages are meaningful (see [this][commit messages] for details)
- [x] Tests have been included and/or updated
- [x] Docstrings have been included and/or updated, as appropriate
- [x] Standalone docs have been updated accordingly
- [x] Changelog(s) has/have been updated, as needed (see `CHANGELOG.md`).


[commit messages]: https://chris.beams.io/posts/git-commit/

 * **Goal**: "Refactor" local scope providers to make implementation and API clearer.
 * **Problem:** `RelativeName` and `ExtRelativeName` are difficult to explain.
 * **Solution:** refactored both scope providers and some internal functions in textx.scoping.tools.py. 
   * Made implementation more reabale:
       * textx.scoping.providers.RelativeName.get_reference_propositions
       * textx.scoping.providers.ExtRelativeName.get_reference_propositions (here, we have much clarity gained!)
   * Especially one accidentally included feature made the code very difficult to understand: When traversing the model with a "model-path", lists were interpreted as list of named-objects. This is not needed by all examples given.
   *  Thus, it is not a 100% refactoring, since we do not allow to have lists within
a path to a desired object (but this was never used in any example). It is
not very probable that this breaks existing code (the old behavior can reactivated with the parameter 'follow_named_element_in_lists' in tools.py with minor differences for postponed references).

goto40 committed:
"Refactored" scope providers to make API clearer.
It is not a 100% refactoring, since we do not allow to have lists within
a path to a desired object (e.g., when extending a class in a DSL with
classes and inheritance - but this was never used in any example). It is
not very probable that this breaks existing code.

goto40 committed:
Added docu for `RelativeName` and `ExtRelativeName`.

goto40 committed:
adapted changelog

goto40 commented:
Note: I extracted this refactoring from ongoing experiments in https://github.com/textX/textX/tree/feature/scope_provider_interactive_demo. This other branch could lead to a proposal how to generate reference name prorpositions (for an editor).

igordejanovic commented on docs/scoping.md:
```diff
@@ -162,7 +162,12 @@ We provide some standard scope providers:
+    - `ExtRelativeName`: The scope is defined by a model-path from the current object and *connected objects*
+        to (typically) a list containing named objects (similar to `RelativeName`). 
+        The *connected objects* are defined again by a model-path from one object 
+        to another named object or a list of named objects.
```
Should these `RelativeName` and `ExtRelativeName` explanations be merged in the proper sections defined up?

BTW, this section on scope providers is not rendered correctly in the browser. You can test the docs locally with `mkdocs serve`.

goto40 replied:
I am sorry for the errors in the rendered version. After corrected the rendering I cleaned up the text. I think it is ok to have no separate section, since with the correctly rendered version you can see all relevant classes on one bullet point depth.

(text adapted)

igordejanovic commented on CHANGELOG.md:
```diff
@@ -35,6 +35,10 @@ please take a look at related PRs and issues and see if the change affects you.
+  - Made scope provider implementation of `RelativeName` and `ExtRelativeName`
+    more readable ([#186]). Minor functional changes, not very probable to have
+    any impact (only affects model-paths containing a list not at the end of the
+    path; see [#186]).
```
If there is a possibility to have any backward incompatibility a changelog entry should be marked with `Possible **(BIC)**`.

goto40 replied:
text added.

igordejanovic commented on tests/functional/examples/test_hierarchical_data_structures_referencing_attributes.py:
```diff
@@ -46,33 +46,43 @@ def test_referencing_attributes():
+                return Postponed()
+            x = get_named_obj_in_list(
+                reference.refs[index - 1].valref.type.vals,
+                attr_ref.obj_name)
```
This whole if/else could be simplified:
```python
base = reference.instance if index == 0 else reference.refs[index - 1]
if base is None or base.type is None:
    return Postponed()
x = get_named_obj_in_list(base.type.vals, attr_ref.obj_name)
```

goto40 replied:
Done.

igordejanovic commented on textx/scoping/providers.py:
```diff
@@ -547,22 +547,41 @@ def __init__(self, path_to_container_object):
+            self.postponed_counter += 1
+            return obj_list
+        assert obj_list is not None
+        assert isinstance(obj_list, list)
```
`get_referenced_object` - the name (and the docstring) doesn't imply that the return value is a list.

goto40 replied:
This is a check for teh correct config of the scope provider.

It is expected to point to a list (of elements).
 * Replaced assert with if+raise
 * Added test
 * Added comment

goto40 replied:
Corrected scoping.md (removed "typically a list" --> "a list")

igordejanovic commented on textx/scoping/providers.py:
```diff
@@ -547,22 +547,41 @@ def __init__(self, path_to_container_object):
+        if type(obj_list) is Postponed:
+            self.postponed_counter += 1
+            return obj_list
+        assert obj_list is not None
```
This assertion could be made at `get_referenced_object` or removed altogether. I don't think it provides some value.

goto40 replied:
Ok. Code removed.

igordejanovic commented on textx/scoping/providers.py:
```diff
@@ -547,22 +547,41 @@ def __init__(self, path_to_container_object):
+        assert obj_list is not None
+        assert isinstance(obj_list, list)
+        for obj in obj_list:
+            assert hasattr(obj, "name")
```
These assertion in the loop are also superfluous as it is check in `filter` which follow (i.e. `AttributeError` will be raised if there is an object without `name`).

goto40 replied:
Ok. code removed.

igordejanovic commented on textx/scoping/providers.py:
```diff
@@ -581,32 +600,51 @@ def __init__(self, path_to_definition_object, path_to_target,
+                self.postponed_counter += 1
+                return def_obj
+
+            tmp_list = get_referenced_object(def_obj, self.path_to_target)
```
Please see the comment on `get_referenced_object` above.

goto40 replied:
(same answer; no extra test added)

igordejanovic commented on textx/scoping/providers.py:
```diff
@@ -581,32 +600,51 @@ def __init__(self, path_to_definition_object, path_to_target,
+            assert tmp_list is not None
+            assert isinstance(tmp_list, list)
+            for obj in tmp_list:
+                assert hasattr(obj, "name")
```
Same comments for assertions as above.

goto40 replied:
code removed.

goto40 replied:
and code changed (check for list); see last comment

igordejanovic commented on textx/scoping/tools.py:
```diff
@@ -130,96 +130,80 @@ def get_recursive_parent_with_typename(obj, desired_parent_typename):
+    if obj is None:
+        return None
+    elif type(obj) is Postponed:
+        return obj
```
Previous 4 lines could be written:
```python
if obj is None or type(obj) is Postponed:
    return obj
```

goto40 replied:
code changed

igordejanovic commented on textx/scoping/tools.py:
```diff
@@ -130,96 +130,80 @@ def get_recursive_parent_with_typename(obj, desired_parent_typename):
+            next_obj = get_named_obj_in_list(obj, names[0])
+        else:
+            raise Exception(
+                "unexpected: got list in path for get_referenced_object")
```
I think it is always better to use `TextXError` (or some subclass of it) for our error so that user of our lib could catch exception at finer granularity.

goto40 replied:
Code changed.

igordejanovic reviewed (changes requested):
Looks good. I like the addition of `get_reference_propositions` as it could be helpful for further tooling development (/cc @danixeee).

There are some changes I propose. Please see the inline comments.

goto40 commented:
> I like the addition of `get_reference_propositions` as it could be helpful for further tooling development (/cc @danixeee).

The additions concerning the editor support will be part of a separate proposition. --> The possible new API will return names, not objects... (to be discussed separately!)

goto40 committed:
fixed docu according to review comments

goto40 committed:
fixed change logs according to review comments

goto40 committed:
simplified code according to review comments

goto40 committed:
commented code to clarify review comments. Added test. replaced assertion with TextXError.

goto40 committed:
removed code according to review comment

goto40 committed:
removed code according to review comment

goto40 committed:
changed code according to review comment

goto40 committed:
changed code according to review comment

goto40 committed:
changed code according to review comment

goto40 committed:
corrected docu (removed "typically")

goto40 commented:
I am done with the corrections defined by the review comments, so far.
I would be happy, if you could have a look at the changes..

igordejanovic commented:
Looks much better. 

There is still an issue with `get_referenced_object`. The name implies that a single object is returned (`object` in the name is singular). The docstrings at the beginning says:
```
get objects based on a path
```
which implies that multiple objects are returned, but in the `Returns:` section:
```
    Returns:
        the object if found, or Postponed() if some postponed
        refs are found on the path / or obj is not found
```
we again have a suggestion that a single object is returned.

Looking at the code it seems that list can be returned also. So this has to be cleared up as it is confusing.

goto40 committed:
renamed function to clarify responsibility. Added unittests. Fixed an error (passed argument to wrong function in tools.py; not catched caused by missing unitests).

goto40 committed:
added tests

goto40 committed:
added docu, corrected function name.

goto40 committed:
flake8

goto40 committed:
Merge branch 'master' into feature/cleanup-localscope

goto40 commented:
(At the end I can squash and merge and rewrite one single proper commit message)

goto40 commented:
You are absolutely right. The function name is misleading and the docu as well. 
 * I adapted the documentation and clarified the responsibility.
 * I renamed the function to a much simple name to "resolve_model_path", because this is what happens.
 * I added some unittests to the existing ones.
   * I found a **bug** I "added" while refactoring: I moved a function argument to the wrong place - and it was covered by the existing tests. **Tests are now adapted** to check this case and the bug is **resolved**.
   * I did not cover the Postponed case (difficult to isolate), but I think most other cases are covered.

igordejanovic commented:
Thanks. It is *much* cleaner now!

> (At the end I can squash and merge and rewrite one single proper commit message)

Yes, please do. Or do interactive rebase and make it in several commits with nice log messages. However you see is better.
